File: processing/syntactic/expressions/tuple_expr.py
from __future__ import annotations

from typing import List

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token import Token
from processing.lexical.tokens.token_type import TokenType
from processing.location import span_marker
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.groups import VarTargetExpression


class TupleExpr(VarTargetExpression):
    @child_property
    def expressions(self) -> List[Expr]:
        pass

    def __init__(
            self,
            parent: Expr = None,
            open_paren: Token = None,
            expressions: List[Expr] = None,
            close_paren: Token = None
    ):
        super().__init__(parent)
        self.open_paren = open_paren
        self.expressions = expressions
        self.close_paren = close_paren
        self.__current_idx = 0

    def __contains__(self, x: object) -> bool:
        return x in self.expressions

    def __len__(self):
        return len(self.expressions)

    def __iter__(self):
        while self.__current_idx < len(self.expressions):
            yield self.expressions[self.__current_idx]
            self.__current_idx += 1

    @span_marker
    def parse_empty(self) -> TupleExpr:
        self.open_paren = self.stream.eat(TokenType.open_parenthesis)
        self.close_paren = self.stream.eat(TokenType.close_parenthesis)
        return self

    def to_axion(self, c: CodeBuilder):
        c += '(', self.expressions, ')'

    def to_csharp(self, c: CodeBuilder):
        c += '(', self.expressions, ')'

    def to_python(self, c: CodeBuilder):
        c += '(', self.expressions, ')'

File: Axion.Core/Processing/Syntactic/Expressions/Generic/TupleExpr.cs
using System.Collections.Generic;
using System.Linq;
using Axion.Core.Processing.CodeGen;
using Axion.Core.Processing.Syntactic.Expressions.TypeNames;
using Axion.Core.Processing.Traversal;
using static Axion.Core.Processing.Lexical.Tokens.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions.Generic {
    /// <summary>
    ///     <c>
    ///         tuple-expr:
    ///             tuple-paren-expr | (multiple-expr [',']);
    ///         tuple-paren-expr:
    ///             '(' multiple-expr [','] ')';
    ///     </c>
    /// </summary>
    public class TupleExpr<T> : Multiple<T> where T : Expr {
        private NodeList<Expr> expressions;

        public NodeList<Expr> Expressions {
            get => expressions;
            set => expressions = BindNode(value);
        }

        [NoTraversePath]
        public override TypeName ValueType => new TupleTypeName(
            this, NodeList<TypeName>.From(
                this,
                Expressions.Select(e => e.ValueType)
            )
        );

        internal TupleExpr(
            Expr               parent,
            IEnumerable<Expr>? expressions = null
        ) : base(parent) {
            Expressions = NodeList<Expr>.From(this, expressions);
            if (Expressions.Count > 0) {
                MarkPosition(
                    Expressions[0],
                    Expressions[^1]
                );
            }
        }

        public TupleExpr<T> ParseEmpty() {
            Stream.Eat(OpenParenthesis);
            Stream.Eat(CloseParenthesis);
            return this;
        }

        public override void ToAxion(CodeWriter c) {
            c.Write("(");
            c.AddJoin(", ", Expressions);
            c.Write(")");
        }

        public override void ToCSharp(CodeWriter c) {
            c.Write("(");
            c.AddJoin(", ", Expressions);
            c.Write(")");
        }

        public override void ToPython(CodeWriter c) {
            c.Write("(");
            c.AddJoin(", ", Expressions);
            c.Write(")");
        }
    }
}